xl2times/transforms.py

```diff
@@ -1475,6 +1511,68 @@ def generate_dummy_processes(
     return tables
 
 
+def process_tradelinks(
+    config: datatypes.Config,
+    tables: List[datatypes.EmbeddedXlTable],
+    model: datatypes.TimesModel,
+) -> List[datatypes.EmbeddedXlTable]:
+    """
+    Transform tradelinks table
+    """
+
+    result = []
+    for table in tables:
+        if table.tag == datatypes.Tag.tradelinks:
+            df = table.dataframe
+            comm = df.columns[0]
+            destinations = list(df.columns).remove(comm)
+            df.rename(columns={comm: "origin"}, inplace=True)
+            df = pd.melt(
+                df, id_vars=["origin"], value_vars=destinations, var_name="destination"
+            )
+            df = df[df["value"] == 1].drop(columns=["value"])
+            df["destination"] = df["destination"].str.upper()
+            df.drop_duplicates(keep="first", inplace=True)
+            df["tradelink"] = 1
+            # Add a column containing linked regions (directionless)
+            df["regions"] = df.apply(
+                lambda row: tuple(sorted([row["origin"], row["destination"]])), axis=1
+            )
+            # Determine whether a trade link is bi- or unidirectional
+            td_type = df.groupby(["regions"])["tradelink"].agg("count").reset_index()
+            td_type.replace({"tradelink": {1: "u", 2: "b"}}, inplace=True)
+            df.drop(columns=["tradelink"], inplace=True)
+            df = df.merge(td_type, how="inner", on="regions")
```

Comment:
This code block above looks good but I do wonder if this sort of thing (melt + count links + create new DF) would actually be faster in plain Python -- it might also be more understandable! It's okay for now I think but I'd be curious to test it sometime. :)

Reply:
I believe we have other instances of code blocks like this. Maybe worthwhile investigating!